Gulp build suite

georgejenkins opened the pull request:
I figured the project could benefit from a modern & extensible build suite. Gulp is almost the de facto  build system for web these days, so that is what I used. It will help people new to the project get up-and-running as the configuration for the build tools are already completed - it now takes only two commands to have Legacy build from source (assuming you have node installed). Most open source web projects provide a similar build system. It also makes build setup very easy and should significantly speed up testing and modification of the source as all changes are immediately compiled and rendered in connected local browsers. 

I've included an npm package for built tool dependencies, and configured gulp to process Legacy's sass, auto-prefix and minify the source, and inject it into all connected browsers via browsersync. I have also updated the readme to reflect the new tools. 

Hit me up on the slack channel if you have any questions/suggestions!

georgejenkins committed:
Created and configured package.json and gulp build suite with sass
preprocessing, minification, and autoprefixing + browsersync for live
changes. TODO: Add source mapping for minified files.

georgejenkins committed:
Changed css link back to dist folder

georgejenkins committed:
Updated README.md to reflect gulp build system

nloveladyallen commented on README.md:
```diff
@@ -19,28 +20,35 @@ Legacy/
+#### Run 'gulp' to start the build suite.
+Once this process has successfully completed, you will be able to run the following command. 
+
+    gulp
```
You have to have the user npm i -g gulp-cli before the gulp command can be used

georgejenkins replied:
Good point. I added a line to the README for this.

georgejenkins committed:
Gulp-cli installation

nloveladyallen commented on gulpfile.js:
```diff
@@ -0,0 +1,42 @@
+// Static Server + watching sass/html files
+gulp.task('serve', ['minify-css'], function() {
+
+    browserSync.init({
```
The rest of the files in the repo use two spaces for indentation, this should be changed to conform to that

nloveladyallen commented on gulpfile.js:
```diff
@@ -0,0 +1,42 @@
+});
+
+gulp.task('minify-css', ['sass'], function() {
+	return gulp.src("./dist/legacy.css")
```
At the very least be consistent within the file.

georgejenkins replied:
No need to be hostile mate, it's just a short oversight.

nloveladyallen replied:
Didn't mean to be hostile, sorry.

georgejenkins committed:
Style update

joexn commented:
Before I merge this, could you give me a demo? Would like to see exactly what it does.

nloveladyallen commented on package.json:
```diff
@@ -0,0 +1,23 @@
+    "url": "https://github.com/joexn/LegacyFramework/issues"
+  },
+  "homepage": "https://github.com/joexn/LegacyFramework#readme",
+  "devDependencies": {
```
If there are dependencies (which there are), they should either be checked in or added to .gitignore.

georgejenkins replied:
In this case, it would be a terrible idea to check in the dependencies - no need to fill the repo with things that are not explicitly Legacy. That is partly why the package exists - so users can resolve dependencies for development if they choose to do so. I do agree on the latter point however, and will push a .gitignore for node_modules next commit.

nloveladyallen commented on gulpfile.js:
```diff
@@ -0,0 +1,39 @@
+var gulp = require('gulp');
+var sass = require('gulp-sass');
+var rename = require('gulp-rename');
+var cleanCSS = require('gulp-clean-css');
```
gulp-clean-css, gulp-autoprefixer, and browser-sync are required here, but not added in package.json

georgejenkins replied:
Good catch! Fixed in next commit.

nloveladyallen commented on gulpfile.js:
```diff
@@ -0,0 +1,39 @@
+
+gulp.task('sass', function() {
+  return gulp.src("./src/legacy.sass")
+  .pipe(sass())
```
This (and the next few) should be indented to line up with `gulp`, i.e.:

``` js
gulp.task('sass', function() {
  return gulp.src("./src/legacy.sass")
             .pipe(sass())
             .on('error', onError)
             .pipe(autoprefixer())
             .pipe(gulp.dest("./dist/"));
});
```

georgejenkins replied:
Hmm. This is more a personal style convention than general js style convention. Gulpfiles typically use one standard indentation past the leftmost return statement. I would prefer to stay consistent with gulp indentation; the change will be reflected in the next commit.

nloveladyallen commented on gulpfile.js:
```diff
@@ -0,0 +1,39 @@
+var autoprefixer = require('gulp-autoprefixer');
+var browserSync = require('browser-sync').create();
+
+gulp.task('default', ['serve']);
```
Why is `serve` the default task? IMO the default should be `minify-css`. Or if you think serve should be default, you should update the README and explain how to only build, not run the server.

georgejenkins replied:
Interesting point. Actually, the default should be `sass` - since the source needs to be processed before minification. `minify-css` only takes the processed css and minifies it. Without `sass` running first, it would simply reminify the same old file. I broke up the two tasks to provide build streaming to both the css and min.css files simultaneously. I'll swap the defaults around & provide info on running browser-sync separately.

nloveladyallen replied:
`minify-css` depends on `sass`, so `minify-css` will both rebuild from source and reminify.

georgejenkins replied:
...yep. You are right. I don't know what I was thinking.

joexn commented:
@georgejenkins @nloveladyallen 

==

As soon as someone demos this to me, I'll merge it.

nloveladyallen commented:
@joexn The next commit will have some improvements, so don't merge yet.

@georgejenkins Can you push the latest commit?

georgejenkins committed:
Updates to package and gulpfile

Resolved some issues with dependencies not being included in the
package. Modified the function of the gulpfile to default to building,
not serving. Updated the readme with new instructions on running gulp.
Added gitignore file to ignore node_modules folder.
